### PR TITLE
feature: Support multi-asset repository scan paths in Semgrep agent

### PR DESCRIPTION
**Pull request by Moonflower1911**
Updates the Semgrep agent to scan the selected repository-like asset directory instead of always scanning the shared `/code` root.

### Changes:
- Derive repository asset scan paths as `/code/<repo_name>_<commit_hash>`.
- Derive repository archive scan paths from the uploaded archive `content_url` UUID `/code/UUID` .
- Rename `REPOSITORY_CODE_PATH` to `ASSETS_CODE_PATH`.
- Add utility helpers for repository and repository archive extraction directory names.

### CODE REVIEW

File: agent/utils.py [hunk 1]
```diff
@@ -40,6 +37,40 @@
 logger = logging.getLogger(__name__)
 
 
+def build_repository_asset_directory(repository_url: str, commit_hash: str) -> str:
```
Comment (burogurama) [inline agent/utils.py]: The name is misleading, it hints that the function creates the directory. Since it derives the name of the directory I suggest `construct_repository_asset_directory_name`

File: agent/utils.py [hunk 2]
```diff
@@ -40,6 +37,40 @@
...
+    if len(repository_path) == 0:
+        repository_path = repository_url
+
+    repository_name: str = os.path.basename(repository_path.rstrip("/"))
+    if repository_name.endswith(".git") is True:
+        repository_name = repository_name[: -len(".git")]
+    return f"{repository_name}_{commit_hash}"
+
+
+def build_repository_archive_asset_directory(content_url: str) -> str:
```
Comment (burogurama) [inline agent/utils.py]: Same here, I suggest `construct_repository_archive_asset_directory_name`

File: agent/utils.py [hunk 3]
```diff
@@ -40,6 +37,40 @@
...
+        Folder name composed from the repository name and commit hash.
+    """
+    parsed_url: parse.ParseResult = parse.urlparse(repository_url)
+    repository_path: str = parsed_url.path
+    if len(repository_path) == 0:
+        repository_path = repository_url
+
+    repository_name: str = os.path.basename(repository_path.rstrip("/"))
+    if repository_name.endswith(".git") is True:
+        repository_name = repository_name[: -len(".git")]
```
Comment (burogurama) [inline agent/utils.py]: Is this necessary, if the code works without I suggest simplifying the implementation.
in build_repository_archive_asset_directory we don t check if the path is empty ot None.
Also I think  such an edge case should not be possible  and we should have failed early.
this is solely for code simplicity.

File: tests/utils_test.py [hunk 4]
```diff
@@ -211,6 +211,51 @@ def testGetFileContent_whenNoContentIsAvailable_shouldReturnNone() -> None:
...
+            "https://github.com/org/repo/",
+            "repo",
+        ),
+    ],
+)
+def testBuildRepositoryAssetDirectory_whenRepositoryUrlHasSuffixes_returnsAssetDirectory(
+    repository_url: str,
+    expected_repository_name: str,
+    repository_commit_hash: str,
+) -> None:
```
Comment (marouaneabeqqal-ostorlab) [inline tests/utils_test.py:231]: doc string as u did before hh

File: tests/utils_test.py [hunk 5]
```diff
@@ -211,6 +211,51 @@ def testGetFileContent_whenNoContentIsAvailable_shouldReturnNone() -> None:
...
+
+
+def testBuildRepositoryArchiveAssetDirectory_whenUploadUrlHasPathAfterId_returnsUploadId() -> (
+    None
+):
+    content_url = "https://example.com/uploads/cc3714/archive/main.zip"
+
+    asset_directory = utils.build_repository_archive_asset_directory(content_url)
+
+    assert asset_directory == "cc3714"
```
Comment (marouaneabeqqal-ostorlab) [inline tests/utils_test.py:276]: same here hh

File: agent/utils.py [hunk 6]
```diff
@@ -1,31 +1,28 @@
...
+import requests
+import tenacity
 from ostorlab.agent.kb import kb
+from ostorlab.agent.message import message as m
 from ostorlab.agent.mixins import (
     agent_report_vulnerability_mixin as vulnerability_mixin,
 )
-from ostorlab.agent.message import message as m
+from ostorlab.assets import android_store, harmonyos_store, ios_store
 from ostorlab.assets import asset as os_asset
```
[the remark this answers is not in the document]
Comment (Moonflower1911) [reply agent/utils.py:22]: new ruff version requires them to be grouped

File: agent/utils.py [hunk 7]
```diff
@@ -40,6 +37,50 @@
...
+    """
+    parsed_url: parse.ParseResult = parse.urlparse(repository_url)
+    repository_path: str = parsed_url.path
+    if len(repository_path) == 0:
+        repository_path = repository_url
+
+    repository_name: str = os.path.basename(repository_path.rstrip("/"))
+    if repository_name.endswith(".git") is True:
+        repository_name = repository_name[: -len(".git")]
+    return f"{repository_name}_{commit_hash}"
```
[the remark this answers is not in the document]
Comment (Moonflower1911) [reply agent/utils.py:56]: we do actually check it, in the function that uses this helper, so no need to do it

File: agent/utils.py [hunk 8]
```diff
@@ -40,6 +37,50 @@
...
+
+
+def build_repository_archive_asset_directory(content_url: str) -> str:
```
[the remark this answers is not in the document]
Comment (Moonflower1911) [reply agent/utils.py]: we do actually check it, in the function that uses this helper, so no need to do it

File: tests/utils_test.py [hunk 9]
```diff
@@ -211,6 +211,51 @@ def testGetFileContent_whenNoContentIsAvailable_shouldReturnNone() -> None:
...
+@pytest.mark.parametrize(
+    ("repository_url", "expected_repository_name"),
+    [
+        (
+            "https://github.com/org/repo.git",
+            "repo",
+        ),
+        (
+            "https://github.com/org/repo/",
+            "repo",
```
[the remark this answers is not in the document]
Comment (Moonflower1911) [reply tests/utils_test.py:223]: these are enough

File: agent/semgrep_agent.py [hunk 10]
```diff
@@ -174,10 +176,35 @@ def process(self, message: m.Message) -> None:
...
+    def _scan_repository_code(
+        self, memory_limit: int, asset_directory: str | None = None
+    ) -> dict[str, Any] | None:
+        """Scan source code extracted to the shared /code volume."""
+        repository_code_path: str = ASSETS_CODE_PATH
+        if asset_directory is not None and len(asset_directory) > 0:
+            if ASSET_DIRECTORY_PATTERN.fullmatch(asset_directory) is None:
+                logger.error(
+                    "Refusing to scan invalid repository asset directory `%s`.",
+                    asset_directory,
```
[the remark this answers is not in the document]
Comment (Moonflower1911) [reply agent/semgrep_agent.py]: your comment is wrong

File: agent/semgrep_agent.py [hunk 11]
```diff
@@ -196,28 +223,53 @@ def _scan_repository_code(self, memory_limit: int) -> dict[str, Any] | None:
 
     def _process_repository_asset(self, message: m.Message, memory_limit: int) -> None:
         """Scan a repository asset and report against its repository URL."""
-        json_output = self._scan_repository_code(memory_limit)
+        repository_url: str | None = message.data.get("repository_url")
+        commit_hash: str | None = message.data.get("commit_hash")
+        asset_directory: str | None = None
+        if repository_url is not None and commit_hash is not None:
```
Comment (yasser-abassi) [inline agent/semgrep_agent.py]: missing repository_url or commit_hash still emits results tagged under this asset, both fields are optional in the proto

File: agent/utils.py [hunk 12]
```diff
@@ -40,6 +37,50 @@
...
+        Path segment immediately after uploads, or the last path segment.
+    """
+    parsed_url: parse.ParseResult = parse.urlparse(content_url)
+    path_segments: list[str] = [
+        segment for segment in parsed_url.path.split("/") if len(segment) > 0
+    ]
+    try:
+        uploads_index: int = path_segments.index("uploads")
+    except ValueError:
+        return os.path.basename(parsed_url.path.rstrip("/"))
```
Comment (yasser-abassi) [inline agent/utils.py]: empty return silently bypasses the guard, scans shared /code unlogged
check is not None only, let "" fail

File: agent/semgrep_agent.py [hunk 13]
```diff
@@ -174,10 +176,35 @@ def process(self, message: m.Message) -> None:
...
                 logger.error("Semgrep completed with errors %s", stderr)
 
-    def _scan_repository_code(self, memory_limit: int) -> dict[str, Any] | None:
-        """Scan the source code extracted to the shared /code volume, the content carried by the message is never read."""
+    def _scan_repository_code(
+        self, memory_limit: int, asset_directory: str | None = None
+    ) -> dict[str, Any] | None:
+        """Scan source code extracted to the shared /code volume."""
+        repository_code_path: str = ASSETS_CODE_PATH
+        if asset_directory is not None and len(asset_directory) > 0:
```
Comment (yasser-abassi) [inline agent/semgrep_agent.py]: asset_directory None skips validation, silently scans full /code instead of refusing like the invalid-name case
is this intentional

File: tests/conftest.py [hunk 14]
```diff
@@ -139,12 +139,18 @@ def scan_message_compressed_js_file() -> message.Message:
 
 
 @pytest.fixture
-def repository_asset_message() -> message.Message:
+def repository_commit_hash() -> str:
```
Comment (yassminassaii) [inline tests/conftest.py:142]: why theres a need to create a function about it if we expect only one str return?

File: agent/utils.py [hunk 15]
```diff
@@ -1,31 +1,28 @@
...
 import dataclasses
-import mimetypes
-import requests
+import json
 import logging
+import mimetypes
 import os
 import re
-from typing import Any, Iterator
+from collections.abc import Iterator
```
Comment (yassminassaii) [inline agent/utils.py:9]: direct import of class

File: tests/utils_test.py [hunk 16]
```diff
@@ -211,6 +211,71 @@ def testGetFileContent_whenNoContentIsAvailable_shouldReturnNone() -> None:
     assert content is None
 
 
+@pytest.mark.parametrize(
```
Comment (1l0c) [inline tests/utils_test.py:214]: Shouldn't we add cases for the other providers as well?

File: agent/utils.py (hunk 15 above)
Comment (Moonflower1911) [reply agent/utils.py:9]: it's permissible for typing-related imports

File: tests/semgrep_agent_test.py [hunk 17]
```diff
@@ -595,23 +610,173 @@ def testAgentSemgrep_whenRepositoryArchiveAssetWithoutContentUrl_emitsBackVulner
...
+        selector="v3.asset.repository",
+        data={"repository_url": "", "commit_hash": repository_commit_hash},
+    )
+
+    test_agent.process(repository_asset_message)
+
+    assert command_mock.call_args.args[0][-1] == semgrep_agent.ASSETS_CODE_PATH
+
+
+def testProcess_whenRepositoryArchiveAssetHasEmptyContentUrl_shouldFallBackToSharedCodePath(
```
Comment (1l0c) [inline tests/semgrep_agent_test.py]: If it's a repository archive, it shouldn't have an empty content_url. We shouldn't fall back in that case; we should raise an error. Falling back could result in scanning the wrong target.

File: agent/utils.py [hunk 18]
```diff
@@ -40,6 +37,48 @@
...
+        Path segment immediately after uploads, or the last path segment.
+    """
+    parsed_url: parse.ParseResult = parse.urlparse(content_url)
+    path_segments: list[str] = [
+        segment for segment in parsed_url.path.split("/") if len(segment) > 0
+    ]
+    try:
+        uploads_index: int = path_segments.index("uploads")
+    except ValueError:
+        return os.path.basename(parsed_url.path.rstrip("/"))
```
Comment (yami-ost) [inline agent/utils.py]: fallback here uses just the basename when there's no uploads segment, that's not guaranteed unique no?

File: tests/utils_test.py [hunk 19]
```diff
@@ -211,6 +211,71 @@ def testGetFileContent_whenNoContentIsAvailable_shouldReturnNone() -> None:
...
+    content_url = "https://example.com/uploads/cc3714?X-Goog-Algorithm=GOO"
+
+    asset_directory = utils.construct_repository_archive_asset_directory_name(
+        content_url
+    )
+
+    assert asset_directory == "cc3714"
+
+
+def testConstructRepositoryArchiveAssetDirectoryName_whenUploadUrlHasPathAfterId_returnsUploadId() -> (
```
Comment (yami-ost) [inline tests/utils_test.py:266]: both tests here have an uploads segment in the url. no test for when uploads is absent

Comment (yami-ost) [inline tests/utils_test.py:266]: nothing tests two archive assets that both fall back to the same basename

File: tests/utils_test.py [hunk 20]
```diff
@@ -211,6 +211,71 @@ def testGetFileContent_whenNoContentIsAvailable_shouldReturnNone() -> None:
...
+) -> None:
+    """Repository URLs with .git or trailing slash use the bare repo name."""
+    asset_directory = utils.construct_repository_asset_directory_name(
+        repository_url, repository_commit_hash
+    )
+
+    assert asset_directory == f"{expected_repository_name}_{repository_commit_hash}"
+
+
+def testConstructRepositoryAssetDirectoryName_whenRepositoryUrlHasNoSuffix_returnsAssetDirectory(
```
Comment (yami-ost) [inline tests/utils_test.py:240]: no test for a repository_url with no path eg just a bare host, produces "_commit_hash" which starts with _ and gets rejected by ASSET_DIRECTORY_PATTERN

File: tests/semgrep_agent_test.py [hunk 21]
```diff
@@ -595,23 +610,173 @@ def testAgentSemgrep_whenRepositoryArchiveAssetWithoutContentUrl_emitsBackVulner
...
+
+    test_agent.process(repository_archive_asset_message)
+
+    expected_scan_path: str = (
+        f"{semgrep_agent.ASSETS_CODE_PATH}/62f54a92-6d5f-4ce8-848e-adf13ff79fee"
+    )
+    assert command_mock.call_args.args[0][-1] == expected_scan_path
+
+
+def testProcess_whenRepositoryAssetDirectoryEscapesAssetsCodePath_shouldNotScan(
```
Comment (yami-ost) [inline tests/semgrep_agent_test.py:676]: this uses a commit_hash with a slash in it so it gets caught by ASSET_DIRECTORY_PATTERN before ever reaching the commonpath check below it, the actual traversal guard has no test that exercises it specifically

File: tests/utils_test.py [hunk 22]
```diff
@@ -211,6 +211,71 @@ def testGetFileContent_whenNoContentIsAvailable_shouldReturnNone() -> None:
...
+    )
+
+    assert asset_directory == "cc3714"
+
+
+def testConstructRepositoryArchiveAssetDirectoryName_whenUploadUrlHasPathAfterId_returnsUploadId() -> (
+    None
+):
+    """Archive content URLs with extra path segments use the upload id."""
+    content_url = "https://example.com/uploads/cc3714/archive/main.zip"
```
Comment (1l0c) [inline tests/utils_test.py:270]: We use GCS, so please make all path and directory name parsing explicit, as we know the expected format. Any deviation should raise an error and be investigated.

Comment (Moonflower1911) [reply tests/utils_test.py:270]: I didn't want to expose the real url we use in gcs, however the essential is that the uuid exists after uploads, I think that's enough

File: agent/semgrep_agent.py [hunk 23]
```diff
@@ -223,15 +223,10 @@ def _scan_repository_code(
...
+        repository_url: str | None = message.data.get("repository_url") or None
+        commit_hash: str | None = message.data.get("commit_hash") or None
         asset_directory: str | None = None
-        if (
-            repository_url is not None
-            and repository_url != ""
-            and commit_hash is not None
-            and commit_hash != ""
-        ):
+        if repository_url is not None and commit_hash is not None:
```
Comment (Moonflower1911) [inline agent/semgrep_agent.py]: keep the check explicit

File: agent/semgrep_agent.py [hunk 24]
```diff
@@ -257,9 +252,9 @@ def _process_repository_archive_asset(
         self, message: m.Message, memory_limit: int
     ) -> None:
         """Scan a repository archive asset and report against its content URL, it carries no repository URL, commit hash nor provider."""
-        content_url: str | None = message.data.get("content_url")
+        content_url: str | None = message.data.get("content_url") or None
         asset_directory: str | None = None
-        if content_url is not None and content_url != "":
+        if content_url is not None:
```
Comment (Moonflower1911) [inline agent/semgrep_agent.py]: same here, keep the check explicit

File: agent/utils.py [hunk 25]
```diff
@@ -40,6 +37,48 @@
...
+        Directory name composed from the repository name and commit hash.
+    """
+    parsed_url: parse.ParseResult = parse.urlparse(repository_url)
+    repository_name: str = os.path.basename(parsed_url.path.rstrip("/"))
+    if repository_name.endswith(".git") is True:
+        repository_name = repository_name[: -len(".git")]
+    return f"{repository_name}_{commit_hash}"
+
+
+def construct_repository_archive_asset_directory_name(content_url: str) -> str:
```
Comment (Moonflower1911) [inline agent/utils.py:59]: instead of falling back `basename`, reject URLs without `/uploads/<uuid>`

File: agent/semgrep_agent.py [hunk 26]
```diff
@@ -174,10 +176,35 @@ def process(self, message: m.Message) -> None:
...
+        """Scan source code extracted to the shared /code volume."""
+        repository_code_path: str = ASSETS_CODE_PATH
+        if asset_directory is not None and len(asset_directory) > 0:
+            if ASSET_DIRECTORY_PATTERN.fullmatch(asset_directory) is None:
+                logger.error(
+                    "Refusing to scan invalid repository asset directory `%s`.",
+                    asset_directory,
+                )
+                return None
+
```
Comment (Oussama-Bakri) [inline agent/semgrep_agent.py]: why we need this matching, we have asset_directory, so just use it, we should not add check to catch values that we don't expect will ever happend, if it happend, that is a bug that need to be fixed

File: tests/utils_test.py (hunk 19 above)
Comment (Moonflower1911) [reply tests/utils_test.py:266]: actually when no uploads is absent we shouldn't fallback to basename

File: agent/semgrep_agent.py (hunk 26 above)
Comment (Moonflower1911) [reply agent/semgrep_agent.py]: even if the values are expected to be valid, as a layer of safety we should check against path traversal or scanning the wrong asset directory...

**Review by 3asm — APPROVED** [review]
